### PR TITLE
Recreation of VideoPlayer.WMS update

### PR DESCRIPTION
**Pull request by KonajuGames**
Recreated from #5084 due to GIT craziness.

### CODE REVIEW

File: MonoGame.Framework/Media/VideoPlayer.WMS.cs [hunk 1]
```diff
@@ -59,102 +85,129 @@ private void PlatformInitialize()
...
-            _clock = null;
+        }
+
+        bool WaitForInternalStateChange(InternalState expectedState, int milliseconds = defaultTimeoutMs)
+        {
+            var timer = System.Diagnostics.Stopwatch.StartNew();
+            while (_internalState != expectedState)
+            {
+#if WINRT
+                Task.Delay(1).Wait();
```
Comment (MichaelDePiazzi) [inline MonoGame.Framework/Media/VideoPlayer.WMS.cs]: I kinda snuck this into the last PR just before you re-submitted it, so you might have missed it...
`Task.Delay` actually generates garbage. An alternative which works just as well and doesn't generate garbage is `Monitor.Wait`. PR #4207 has an example of this (specifically commit 0a18376).

**Comment by KonajuGames** [comment]
Thanks @MichaelDePiazzi.  I did see that on the old PR, and forgot it when remaking this PR.  I've added that change.

**Comment by Jjagg** [comment]
image


I just tried this and got this exception. Seems some dependencies do not come with the Windows 10 installation :/

**Comment by KonajuGames** [comment]
:rage: I seem to recall seeing a similar exception a long time back.  Not sure what it was that I installed to fix it.

**Comment by KonajuGames** [comment]
@Jjagg Will you be able to try this branch again?  I couldn't repro the issue, but I did find the previous reference to this same issue (issue #2620 and PR #3255).  I have now added that fix to it.

**Comment by Jjagg** [comment]
The change did not help. It's just failing after the 10 retries now.

image

Let me know if I can do anything to help.

**Comment by KonajuGames** [comment]
Windows is great and all, but some of their APIs are just garbage.  I realised I'd have to fix the time delay in that loop anyway for UWP.

Is it one video or all videos you try that fail?  Can you try the videos in the unit tests?

**Comment by Jjagg** [comment]
The video in the unit tests works!
The video that fails is one I had lying around on my laptop (mp4).

**Comment by KonajuGames** [comment]
Could you make that video available somewhere so I can test with it?

**Comment by Jjagg** [comment]
I uploaded it to Dropbox:
https://www.dropbox.com/s/4lv6lf5z2k518hu/atrium.mp4?dl=0

It's a video that came with the MATLAB installation.

**Comment by KonajuGames** [comment]
Ahh, that video has no audio tracks at all.  I've allowed it to fail when getting the video controller, but now the sample grabber is not returning any texture data.

**Comment by romanov** [comment]
Any news?

**Comment by KonajuGames** [comment]
I will get back onto this on Monday.  I've been away for a few weeks and
busy at work.​

**Comment by romanov** [comment]
@KonajuGames i've just checked the development build and looks like this: https://github.com/MonoGame/MonoGame/pull/5737 fixes the issue with memory leaks (i've tested .mp4 file)

**Comment by KonajuGames** [comment]
Re-based to current develop branch, and fixed the resuming.  It was throwing an invalid pointer error.  Works fine with a single `VideoPlayer` instance.  Multiple concurrent `VideoPlayer` instances will show the same video,  Haven't worked that one out yet.

**Comment by gap9269** [comment]
Is there any plan to get this merged soon (particularly the Resume() bug that causes a crash)?

**Comment by SeriousMartin** [comment]
Works great for me, with one addition:
In VideoPlayer.WMS.CreateTexture add a check to _texture.IdDisposed. For some reason I couldn't inspect, the texture was disposed and CreateTexture never recreated it, because the null check returns false.

**Comment by SeriousMartin** [comment]
@KonajuGames @tomspilman 
Any chance to get this merged (including my adjustment) soon? Can I help in any way?

**Comment by simon-riley** [comment]
Is there any updates on the status of this being merged?

**Comment by MichaelDePiazzi** [comment]
I recently started using a video for the background of my games menu screen. So I pulled in this PR to my fork so I could take advantage of the improvements and fixes. However, while I was testing this I did find a couple of issues with the video player.

The first issue was a memory leak after disposing the video player and/or video. So I did some investigation and found this article in particular to be very useful:
https://docs.microsoft.com/en-us/windows/desktop/medfound/step-7--shut-down-the-media-session

And here's my commit which attempts to shut everything down properly:
https://github.com/MichaelDePiazzi/MonoGame/commit/652be335729fd99ea340d3adfd373f69d91e4fad

But unfortunately even this did not seem to get rid of the memory leak! So I'm at a bit of a loss as to why it's happening... Maybe it might be a bug in SharpDX?

Anyway, to work around it I ended up using a static instance of the video player and video. Not ideal, but at least it minimises the leak.

I also had some playback issues when trying to play the same video again (e.g. stop the video and then play it again later, but the video doesn't start). This was a similar problem to what we had with song playback, so the fix was basically the same. Here's my commit which fixes this:
https://github.com/MichaelDePiazzi/MonoGame/commit/e4c96141200e341cf94b092f5a11d697e66d6bb3

I did one more minor commit related to this. Occasionally when disposing the video player, I got an exception because an event came in after the session was disposed for some reason. So I tried to make this a little more robust:
https://github.com/MichaelDePiazzi/MonoGame/commit/c6aa30fa90cdc231d8747e9371d10efec024c869

Anyway, I hope this is useful.

I plan to do some further testing on this still. And I will be releasing an experimental build of my game soon to put this through it's paces some more.

I would love to see this PR merged sometime soon as the video player definitely has some issues right now. I'm more than happy to submit my additional changes as a separate PR, but this PR should probably be merged first. Or feel free to just include my changes with this PR if you want.

If there's anything more I can do to help with this, please let me know.

(ping @KonajuGames @tomspilman @Jjagg)

**Comment by SeriousMartin** [comment]
@KonajuGames Will you add the suggestions of @MichaelDePiazzi to this pull requests? Anything I can do to help?

**Comment by KonajuGames** [comment]
@SeriousMartin @MichaelDePiazzi I decided to give this another go.  Thanks for your suggestions and input.  It lead me to solve some of the issues it had with playing multiple videos at the same time.  I just have to sort out the 10MB unreleased memory allocation when playing a video and 3MB unreleased memory allocation when resuming a paused video.

**Comment by MichaelDePiazzi** [comment]
@KonajuGames That's great to hear! 😃
Please let me know if I can help with anything at all (e.g. testing, code review). I'm pretty flat out right now (releasing Final Days in 4 weeks! 😅), but I'd really like to see the video player sorted out.

**Comment by tomspilman** [comment]
We will be moving to a native video player in the near future.

**Pull request closed without merging**